File: testjava/src/main/java/com/arrays/MoveAllZeros.java
package com.arrays;

import java.util.Arrays;

public class MoveAllZeros {

    public static void main(String[] arg) {
        int[] arr = new int[]{0,1,0,3,12,0};
        moveAllZeros(arr);
        System.out.println();
        moveAllZeros2(arr);
    }

    private static void moveAllZeros2(int[] arr) {
        int writer = 0;
        int reader = 0;
        System.out.println("moveAllZeros2 Array before move: "+ Arrays.toString(arr));
        while (reader < arr.length){
            if (arr[reader] != 0){
                arr[writer] = arr[reader];
                writer++;
            }
            reader++;
        }
        while (writer < arr.length){
            arr[writer] = 0;
            writer++;

        }
        System.out.println("moveAllZeros2 Array after move: "+Arrays.toString(arr));
    }

    private static void moveAllZeros(int[] arr) {
        System.out.println("Array before move: "+ Arrays.toString(arr));
        int[] newArr = new int[arr.length];
        int count = 0;
        for (int i = 0; i < arr.length; i++) {
            if (arr[i] != 0){
                newArr[count] = arr[i];
                count++;
            }
        }
        for (int i = count; i >0 ; i--) {
            newArr[count] = 0;
        }

        System.out.println("Array after move: "+Arrays.toString(newArr));
    }
}

File: kotlin_prog/src/main/java/com/zafar/kotlin/arrays/MoveAllZeros.kt
package com.zafar.kotlin.arrays

fun main(arg: Array<String>) {
    val intArray = intArrayOf(0,1,0,3,12,0) // input
    // output should be {1,3,12,0,0,0}
    moveAllZeros(intArray)
    println()
    moveAllZeros1(intArray)
}

fun moveAllZeros1(arr: IntArray) {
    println("While Before moving Array: ${arr.contentToString()}")
    var reader = 0
    var writer = 0
    while (reader < arr.size) {
        if (arr[reader] != 0){
            arr[writer] = arr[reader]
            writer++
        }
        reader++
    }

    while (writer < arr.size) {
        arr[writer] = 0
        writer++
    }
    println("While After moving Array: ${arr.contentToString()}")
}

fun moveAllZeros(intArray: IntArray) {
    println("Before moving Array: ${intArray.contentToString()}")
    var count = 0
    val newArray = IntArray(intArray.size)
    for (i in intArray.indices) {
        if (intArray[i] != 0){
            newArray[count] = intArray[i]
            count++
        }
    }

    for (i in count downTo 1){
        newArray[count] = 0
    }
    println("After moving Array: ${newArray.contentToString()}")
}
